icicle-compiler/src/Icicle/Sea/Header.hs

```diff
@@ -0,0 +1,557 @@
+        (StringT <$) . pUnit
+
+    "unit" ->
+      Just $
+        (UnitT <$) . pUnit
+
+    "error" ->
+      Just $
+        (ErrorT <$) . pUnit
+
+    "fact_identifier" ->
+      Just $
+        (FactIdentifierT <$) . pUnit
+
+    "array" ->
+      Just $
+        fmap ArrayT . pValType
+
+    "map" ->
+      Just . Aeson.withObject "MapT" $ \o ->
+        MapT
+          <$> withKey "key" o pValType
+          <*> withKey "value" o pValType
+
+    "option" ->
+      Just $
+        fmap OptionT . pValType
+
+    "pair" ->
+      Just . Aeson.withObject "PairT" $ \o ->
+        PairT
+          <$> withKey "first" o pValType
+          <*> withKey "second" o pValType
+
+    "sum" ->
+      Just . Aeson.withObject "SumT" $ \o ->
+        SumT
+          <$> withKey "left" o pValType
+          <*> withKey "right" o pValType
+
+    "struct" ->
+      Just $
+        fmap StructT . pStructType
+
+    "buf" ->
+      Just . Aeson.withObject "BufT" $ \o ->
+        BufT
+          <$> withKey "size" o pInt
+          <*> withKey "element" o pValType
+
+    x ->
+      fail $ "Unknown type: " <> Text.unpack x
+
+ppValType :: ValType -> Aeson.Value
+ppValType = \case
+ BoolT ->
+   ppEnum "bool" ppUnit
+
+ TimeT ->
+   ppEnum "time" ppUnit
+
+ DoubleT ->
+   ppEnum "double" ppUnit
+
+ IntT ->
+   ppEnum "int" ppUnit
+
+ StringT ->
+   ppEnum "string" ppUnit
+
+ UnitT ->
+   ppEnum "unit" ppUnit
+
+ ErrorT ->
+   ppEnum "error" ppUnit
+
+ FactIdentifierT ->
+   ppEnum "fact_identifier" ppUnit
+
+ ArrayT x ->
+   ppEnum "array" $
+     ppValType x
+
+ MapT k v ->
+   ppEnum "map" $
+     Aeson.object [
+         "key" .=
+           ppValType k
+       , "value" .=
+           ppValType v
+       ]
+
+ OptionT x ->
+   ppEnum "option" $
+     ppValType x
+
+ PairT x y ->
+   ppEnum "pair" $
+     Aeson.object [
+         "first" .=
+           ppValType x
+       , "second" .=
+           ppValType y
+       ]
+
+ SumT x y ->
+   ppEnum "sum" $
+     Aeson.object [
+         "left" .=
+           ppValType x
+       , "right" .=
+           ppValType y
+       ]
+
+ StructT x ->
+   ppEnum "struct" $
+     ppStructType x
+
+ BufT n x ->
+   ppEnum "buf" $
+     Aeson.object [
+         "size" .=
+           ppInt n
+       , "element" .=
+           ppValType x
+       ]
+
+pMeltedType :: Aeson.Value -> Aeson.Parser MeltedType
+pMeltedType =
+  Aeson.withObject "MeltedType" $ \o -> do
+    logical <- withKey "logical" o pValType
+    melted0 <- withKey "melted" o $ pList "Array ValType" pValType
+
+    let
+      melted =
+        meltType logical
+
+    when (melted0 /= melted) $
+      fail $
+        "serialised melted type did not match calculated melted type:" <>
+        "\n" <>
+        "\n  logical type =" <>
+        "\n    " <> show logical <>
+        "\n" <>
+        "\n  serialised melted type =" <>
+        "\n    " <> show melted0 <>
+        "\n" <>
+        "\n  calculated melted type =" <>
+        "\n    " <> show melted
```

Comment:
Yeah it's `'impossible'` in the GHC sense, it would happen if the melting algorithm changed, I figured there is no harm checking it.

The alternative is to never store the melted types are just regenerate them.